Feature/orange 795 consolidated logging

puckybreg opened the pull request:
If this PR fixes a bug, you _must_ add test cases representative of the bug.
#### What's this PR do?
Adds a formatter to our winston logger which is approaching the JSON logging format defined in the Orange-795 ticket

### What is this PR still lacking? 
- The logging message that lets the dev know the service has started get's sent to our winston log before the async call to get the last git commit info comes back and thus, this initial log message does not get the consolidated logging format. 
- I am still unclear on the `event` subsection of the logging format
- I'm having trouble getting the error stack-trace out of the info object provided by `printf`

#### Related JIRA tickets:
Orange-795  https://jira.amida.com/browse/ORANGE-795
#### How should this be manually tested?
Do a yarn install and a `yarn start` and hit any of the messaging endpoints
#### Any background context you want to provide?
#### Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/8662143/47669944-66b26980-db82-11e8-8ab9-f4036dfb97c5.png)

puckybreg committed:
WIP on consolidated logging

Commit:
add user create end point for user table

Commit:
remove ssl migration from new branch

puckybreg committed:
Adding git commit info to logging format

puckybreg committed:
Finalized logging format

puckybreg committed:
Merge branch release/2.3.0 into feature/Orange-795-Consolidated-Logging

puckybreg committed:
Merge branch release/2.3.0 into feature/Orange-795-Consolidated-Logging take 2

puckybreg committed:
remove test logging statement

puckybreg committed:
add in err object

Commit:
add package.json changes for unit tests

rmharrison commented:
Proximal cause of test failure
```
Database connection failed: { error: password authentication failed for user "amida"
```

morleegan reviewed (approved):
Looks good overall other than the comments you left in your original PR. 
Also, it might be good to add `prettyPrint()` into the formatting of at least  `production`? I would think that would make it easier to understand.

rmharrison commented:
`prettyPrint()` for dev, fine. Prod should be straight JSON for consumption downstream.

Once https://github.com/eugeny-dementev/winston-console-formatter/pull/14 is merged down, we can use that library for nice printing in dev.

rmharrison commented on config/winston.js:
```diff
@@ -13,12 +17,49 @@ const logger = createLogger({
+        meta: {
+            service: {
+                commit: commit.shortHash,
+                commitMessage: commit.subject,
```
I wouldn't include the `commit.subject`. `shortHash` is sufficient.

puckybreg replied:
got it

rmharrison commented on config/winston.js:
```diff
@@ -13,12 +17,49 @@ const logger = createLogger({
 
-if (config.env !== 'production') {
-    logger.format = combine(
+git.getLastCommit((err, commit) => {
```
## Question: Removing git binary dependency?
`git-last-commit` uses a git binary under the hood, see: https://github.com/seymen/git-last-commit/blob/master/source/index.js#L30

Mike's proposal is to build a git info and read from that, to avoid a git binary dep in prod.

Usecase for`npm:husky`

Naturally, the downside is that we'd need to include the git info at the build step. For [DEVOPS-343](https://jira.amida.com/browse/DEVOPS-343), I don't know how one does that with just an npm install of the formatter.
Build step won't be done in all repos, so please fail gently, i.e. don't include the commitHash if the file isn't there. Should work with or without.

mike [29 Oct 2018, 5:08 PM]
You mentioned something to retrieve git info for reference in a release, what was it?  I created a one liner and just made it part of the build process but am open to a more formal solution.  The one thing I wanted to avoid if possible is a dependency on git in the production release.

mike [29 Oct 2018, 5:28 PM]
my quick and dirty solution.  I did not want to spend too much time on it but wanted something reproducible.  In the long run I would probably tweak the output a bit or parse raw.
Untitled 
  "build": "babel src --out-dir dist && git show --format='{\"hash\":\"%H\",\"committer\":\"%cn\",\"commitDate\":\"%cD\",\"subject\":\"%s\"}' --no-patch >dist/git.info"

## The first line upon `npm start` doesn't have the custom formatting

```
$ npm start
...
{"service":"messaging-service","message":"server started on port","port":4001,"node_env":"development","level":"info"}
```

https://github.com/amida-tech/amida-messaging-microservice/blob/00629319f9ff78911d9acb3d949d06cb9c0857b9/index.js#L17

May be cause because the logger ([`export logger`](https://github.com/amida-tech/amida-messaging-microservice/blob/00629319f9ff78911d9acb3d949d06cb9c0857b9/config/winston.js#L65) ) is being used here **before** the callback(commit) completes. Using Mike's solution (synchronous file read) should resolve.

rmharrison replied:
Answer: Both resolved by removing git-last-commit dep. Commit not included in log message.

rmharrison commented on config/winston.js:
```diff
@@ -13,12 +17,49 @@ const logger = createLogger({
+git.getLastCommit((err, commit) => {
+    const prodFormat = printf(info => JSON.stringify({
+        service: pjson.name,
+        logger: 'my_winston_logger',
```
2 cents: Would prefer something like "application_logger"

rmharrison commented on config/winston.js:
```diff
@@ -13,12 +17,49 @@ const logger = createLogger({
+                time: info.timestamp,
+                baz: 'bok',
+            },
+            event: {
```
## Optional
Even with the formater
```
logger.info("hello", {'foo': bar})
```

Would yield
```
{
  msg: 'hello'
  foo: 'bar'
}
```
Do you know how easy it'd be to have it go into "event"?
```

{
  msg: 'hello'
  event: {
     foo: 'bar'
  }
}
```

rmharrison replied:
Done.

rmharrison reviewed (commented):
Boom!

Run: `docker-compose -f docker-compose.yml up`
Hit: http://localhost:4001/api/health-check

```
$ 
{
  "service": "amida-messaging-microservice",
  "logger": "my_winston_logger",
  "hostname": "f5171444ad70",
  "level": "info",
  "msg": "HTTP GET \/api\/health-check 304 9ms",
  "meta": {
    "service": {
      "commit": "0062931",
      "commitMessage": "add in err object",
      "version": "1.0.0"
    },
    "logger": {
      "time": "2018-10-31T04:56:45.726Z",
      "baz": "bok"
    },
    "event": {
      "foo": "bar"
    }
  },
  "err": {
    
  }
}
```

Almost there!

Commit:
add command for jenkins tests and handle for deletion of database does not exist

nickmccally committed:
Merge pull request #74 from amida-tech/feature/add-uuid-v2

Feature/add UUID v2

rmharrison commented:
@puckybreg : Will you have time to get this merged down this week? 
- Dump git commit hash in build step instead of using git binary (suspect fixing will also solve first log line doesn't have formatting), https://github.com/amida-tech/amida-messaging-microservice/pull/73#discussion_r229560517
- Fix Jenkins error (w/ @nickmccally?)

puckybreg committed:
Easy Wins from PR comments

rmharrison commented:
From @nickmccally : Merge develop to fix Jenkins failure.

Commit:
Hotfix: add port to database.js for sequelizeCLI

rmharrison committed:
Merge branch 'develop' into branch 'feature/Orange-795-Consolidated-Logging'

Includes jenkins fix (tests were failing) and `yarn migrate` fix

rmharrison committed:
[ORANGE-795] Remove `git-last-commit` dependency

Including the commit message via `git-last-commit` introduces a dependency for `git` in production.
`git` is *not* installed in light-weight base images such as [node:10-alpine](https://github.com/nodejs/docker-node/blob/526c6e618300bdda0da4b3159df682cae83e14aa/10/alpine/Dockerfile).

rmharrison committed:
[ORANGE-795] Log-line specific meta data under 'meta.event'

`logger.info('hello', {foo: 'bar'})`

Previously:
```
{
  foo: 'bar'
  msg: 'hello'
  meta: {
    event: {}
  }
}
```

Now:
```
{
  msg: 'hello'
  meta: {
    event: {
      foo: 'bar'
    }
  }
}
```

rmharrison commented:
Just noticed the PR was created against `branch:release-2.3.0` and not `branch:develop`.

As a feature, it *should* have gone into `branch:develop`. That said, causes more confusion to fix.
